module/zfs/zfs_vfsops.c

```diff
@@ -1415,6 +1415,21 @@ zfs_preumount(struct super_block *sb)
 
 	if (zsb)
 		zfsctl_destroy(sb->s_fs_info);
+	/*
+	 * wait for iput_async before entering evict_inodes in
+	 * generic_shutdown_super. The reason we must finish before
+	 * evict_inodes is iput when lazytime and zfs_purgedir calling
+	 * zfs_zget would bump i_count from 0 to 1. This would race with the
+	 * i_count check in evict_inodes, so it could destroy the inode while
+	 * we are still using it.
+	 *
+	 * We wait for two pass, first pass are xattr dirs, second pass are
+	 * xattr entries.
+	 */
+	taskq_wait_outstanding(dsl_pool_iput_taskq(
+	    dmu_objset_pool(zsb->z_os)), 0);
+	taskq_wait_outstanding(dsl_pool_iput_taskq(
+	    dmu_objset_pool(zsb->z_os)), 0);
```

Comment:
If someone does zfs_purgedir in the first pass, it will add xattr entries into the taskq. That's what I meant by first pass are xattr dirs, second pass are xattr entries. I'll update the comment to be more clear on this.